### PR TITLE
Multiview painting: Studio

### PR DESCRIPTION
**Pull request by dkwingsmt**
*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/ui/window.dart [hunk 1]
```diff
@@ -289,10 +289,10 @@ class FlutterView {
   ///   scheduling of frames.
   /// * [RendererBinding], the Flutter framework class which manages layout and
   ///   painting.
-  void render(Scene scene) => _render(scene);
+  void render(Scene scene) => _render(viewId == 0 ? 0 : viewId.hashCode, scene);
```
Comment (loic-sharma) [inline lib/ui/window.dart:292]: Why `viewId.hashCode`? Should this be `viewId as int`?
cene);

Comment (dkwingsmt) [reply lib/ui/window.dart:292]: Good point, it was a panic move because I didn't know how to convert Object to int...

File: flow/surface.h [hunk 2]
```diff
@@ -35,22 +35,13 @@ class Surface {
 
   virtual bool IsValid() = 0;
 
-  virtual std::unique_ptr<SurfaceFrame> AcquireFrame(const SkISize& size) = 0;
+  virtual std::unique_ptr<SurfaceFrame> AcquireFrame(int64_t view_id,
```
Comment (loic-sharma) [inline flow/surface.h:38]: Could you explain why this has a `view_id` parameter? We will have separate surfaces for each view right?

File: flow/surface.h [hunk 3]
```diff
@@ -35,22 +35,13 @@ class Surface {
...
+                                                     const SkISize& size) = 0;
 
   virtual SkMatrix GetRootTransformation() const = 0;
 
   virtual GrDirectContext* GetContext() = 0;
 
-  virtual std::unique_ptr<GLContextResult> MakeRenderContextCurrent();
-
-  virtual bool ClearRenderContext();
```
Comment (loic-sharma) [inline flow/surface.h:46]: Could you explain why you removed the context methods from the surface?

Comment (dkwingsmt) [reply flow/surface.h:46]: Because they don't seem needed :)  I was actually testing whether they were, and after I removed these methods it actually compiles.

File: shell/common/rasterizer.cc [hunk 4]
```diff
@@ -125,51 +124,89 @@ void Rasterizer::DisableThreadMergerIfNeeded() {
...
-    return;
+int Rasterizer::DrawLastLayerTree(
+    std::unique_ptr<FrameTimingsRecorder> frame_timings_recorder,
+    bool enable_leaf_layer_tracing) {
+  if (!studio_) {
+    return 0;
+  }
+  int success_count = 0;
+  bool should_resubmit_frame = false;
+  for (auto& record_pair : surfaces_) {
```
Comment (loic-sharma) [inline shell/common/rasterizer.cc:184]: Should this be able to render the last layer tree for just a single surface?

From my understanding, `DrawLastLayerTree` is used in scenarios where an external texture was updated but the framework's layer tree is otherwise unchanged. For example, if a video player needs to update the surface on each video frame even though the Flutter app itself hasn't rendered a frame. We shouldn't re-draw a second window that doesn't contain the video.

File: shell/common/rasterizer.cc [hunk 5]
```diff
@@ -185,17 +222,19 @@ RasterStatus Rasterizer::Draw(
...
+  DoDrawResult draw_result;
   LayerTreePipeline::Consumer consumer =
-      [&](std::unique_ptr<LayerTreeItem> item) {
+      [this, &draw_result,
+       &discard_callback](std::unique_ptr<LayerTreeItem> item) {
         std::shared_ptr<LayerTree> layer_tree = std::move(item->layer_tree);
         std::unique_ptr<FrameTimingsRecorder> frame_timings_recorder =
             std::move(item->frame_timings_recorder);
+        int64_t view_id = item->view_id;
         if (discard_callback(*layer_tree.get())) {
```
Comment (loic-sharma) [inline shell/common/rasterizer.cc:233]: I believe this discard callback is used when a window was resized. We discard the scene if its dimensions don't match the window's new size. We'll need to update this callback as each surface may have different dimensions.

File: flow/surface.h (hunk 2 above)
Comment (dkwingsmt) [reply flow/surface.h:38]: Yeah you are right.

Initially I added the view_id to `SurfaceFrame::Submit`, so that surfaces and surface frames can be "view agnostic" until the last second so that they feel more "pure", but it turns out that we want frame buffers to be tied with a view ID so that buffers can be invalidated more easily. I moved `view_id` earlier to `AcquireFrame`, but on a second thought we should just straight up tie surfaces with views anyway.

File: shell/common/rasterizer.cc (hunk 4 above)
Comment (dkwingsmt) [reply shell/common/rasterizer.cc:184]: If there are two videos in the app on two different windows, then we'll need the last-frame information for both views.

File: shell/common/rasterizer.cc (hunk 5 above)
Comment (dkwingsmt) [reply shell/common/rasterizer.cc:233]: Sounds reasonable, I'll update it! Thank you.

File: flow/surface.h (hunk 2 above)
Comment (dkwingsmt) [reply flow/surface.h:38]: Although I'm wondering if we can assume that all surfaces for a view are of the same size, so that if we invalidate the frame buffer for one surface of a view, then we can also invalidate all other frame buffers of that view.

File: shell/common/rasterizer.cc (hunk 4 above)
Comment (loic-sharma) [reply shell/common/rasterizer.cc:184]: Right, but do we need to rasterize all windows if only one window has a video player?

File: flow/surface.h (hunk 2 above)
Comment (loic-sharma) [reply flow/surface.h:38]: > I'm wondering if we can assume that all surfaces for a view are of the same size

Are you including the macOS embedder's surfaces here? If so, I don't think so: the engine may have one layer for a button and another layer for the app's background. From my understanding, these would translate into macOS embedder surfaces with different dimensions.

File: shell/common/rasterizer.cc (hunk 4 above)
Comment (dkwingsmt) [reply shell/common/rasterizer.cc:184]: I think you have a valid point, but let me rephrase your question.

In the case that multiple views have a video player each, we should probably allow `DrawLastLayerTree` to only draw the views that need a resubmission, i.e. whose video player triggered this vsync - instead of resubmitting all views on all vsyncs.

However, this feature is tricky. We need to allow VSyncs to carry information on the views that triggered it, so that the animator can send it back to the rasterizer.

Also, this feature can be done as a separate issue after the main changes, since it affects performance, not functionality.
